Unable to create PR. already exists; Delete the existing branch and try again.

**Describe the bug**
After the first execution i get this kind of errors: `Unable to create pull request 'Bump @testing-library/user-event from 13.5.0 to 14.6.1 in /frontend' as source branch 'dependabot/npm/main/frontend/testing-library/user-event-14.6.1' already exists; Delete the existing branch and try again.`

**Categorization**
- [x] This is not a permissions issue (Seek help at https://github.com/tinglesoftware/dependabot-azure-devops/discussions/1245)

**Repository**
URL: e.g. https://dev.azure.com/tingle/dependabot/_git/repro-684
<!-- If credentials are required, send them to support@tingle.software -->

**To Reproduce**
Steps to reproduce the behavior:
1. I've created this pipeline:

```

trigger: none # Disable CI trigger

schedules:
- cron: '0 0 * * *' # weekly on sunday at midnight UTC
  always: true # run even when there are no code changes
  branches:
    include:
      - main
  batch: true
  displayName: Weekly

pool:
  vmImage: 'ubuntu-latest' # requires macos or ubuntu (windows is not supported)

steps:
- task: dependabot@2
  inputs:
    mergeStrategy: 'squash'
```


2. I have all permissions mentioned in the discussion thread.

**Expected behavior**

If the pipeline is executed several time and the PR are still active I shouldn't get an error.

**Logs and screenshots**
-

**Extension (please complete the following information):**
 - Host: Azure DevOps

**Server (please complete the following information):**
 - Region westeurope
 
**Additional context**
-

> [@0GiS0](https://github.com/0GiS0) are you able to attach the pipeline logs for the Dependabot task?

Thank you so much for your answer! I tried but the log is too big and I cannot paste in! 

This is the last piece:

````
2025-02-24T07:29:06.3686584Z ##[debug]🌎 🠊 [GET] https://***/***/***/_apis/git/repositories/***?api-version=5.0
2025-02-24T07:29:06.3687127Z ##[debug]🌎 🠈 [200] OK
2025-02-24T07:29:06.3688442Z ##[error]Unable to create pull request 'Bump web-vitals from 2.1.4 to 4.2.4 in /frontend' as source branch 'dependabot/npm/main/frontend/web-vitals-4.2.4' already exists; Delete the existing branch and try again.
2025-02-24T07:29:06.3689461Z ##[debug]Processed: ##vso[task.issue type=error;source=TaskInternal;]Unable to create pull request 'Bump web-vitals from 2.1.4 to 4.2.4 in /frontend' as source branch 'dependabot/npm/main/frontend/web-vitals-4.2.4' already exists; Delete the existing branch and try again.
2025-02-24T07:29:06.3689951Z ##[section]Processing 'mark_as_processed'
2025-02-24T07:29:06.3690323Z ##[debug]{"base-commit-sha":"64fd5ca8da1c0e0cbea516e3dbf9a048b30c18be"}
2025-02-24T07:29:06.3690600Z ##[endgroup]
2025-02-24T07:29:06.3690901Z ##[debug]task result: Failed
2025-02-24T07:29:06.3691538Z ##[error]3 update tasks(s) failed, check logs for more information
2025-02-24T07:29:06.3692118Z ##[debug]Processed: ##vso[task.issue type=error;source=TaskInternal;]3 update tasks(s) failed, check logs for more information
2025-02-24T07:29:06.3703150Z ##[debug]Processed: ##vso[task.complete result=Failed;]3 update tasks(s) failed, check logs for more information
2025-02-24T07:29:06.3715659Z ##[section]Finishing: dependabot
````

--- a/extension/tasks/dependabotV2/index.test.ts
+++ b/extension/tasks/dependabotV2/index.test.ts
@@ -1,0 +1,141 @@
+import { performDependabotUpdatesAsync } from './index';
+import { IPullRequestProperties } from './utils/azure-devops/interfaces/IPullRequest';
+import { DependabotCli } from './utils/dependabot-cli/DependabotCli';
+import { DependabotJobBuilder } from './utils/dependabot-cli/DependabotJobBuilder';
+import { DependabotOutputProcessor } from './utils/dependabot-cli/DependabotOutputProcessor';
+import { IDependabotUpdateOperationResult } from './utils/dependabot-cli/interfaces/IDependabotUpdateOperationResult';
+import { IDependabotConfig } from './utils/dependabot/interfaces/IDependabotConfig';
+import { ISharedVariables } from './utils/getSharedVariables';
+import { GitHubGraphClient } from './utils/github/GitHubGraphClient';
+
+jest.mock('./utils/dependabot-cli/DependabotCli');
+jest.mock('./utils/dependabot-cli/DependabotJobBuilder');
+jest.mock('./utils/github/GitHubGraphClient');
+
+const tsDependabotJobBuilder = require('./utils/dependabot-cli/DependabotJobBuilder');
+const tsDependabotOutputProcessor = require('./utils/dependabot-cli/DependabotOutputProcessor');
+
+describe('performDependabotUpdatesAsync', () => {
+  let taskInputs: ISharedVariables;
+  let dependabotConfig: IDependabotConfig;
+  let dependabotCli: DependabotCli;
+  let dependabotCliUpdateOptions: any;
+  let existingPullRequests: IPullRequestProperties[];
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    taskInputs = {} as ISharedVariables;
+    dependabotConfig = {
+      updates: [
+        {
+          'package-ecosystem': 'npm',
+          'directory': '/',
+        },
+      ],
+      registries: {},
+    } as IDependabotConfig;
+    dependabotCli = new DependabotCli(DependabotCli.CLI_PACKAGE_LATEST, null, true);
+    dependabotCli.update = jest
+      .fn()
+      .mockResolvedValue([
+        { success: true, output: { type: 'mark_as_processed', data: {} } },
+      ] as IDependabotUpdateOperationResult[]);
+    dependabotCliUpdateOptions = {};
+    existingPullRequests = [];
+  });
+
+  it('should perform "update all" job successfully', async () => {
+    const failedUpdates = await performDependabotUpdatesAsync(
+      taskInputs,
+      dependabotConfig,
+      dependabotConfig.updates,
+      dependabotCli,
+      dependabotCliUpdateOptions,
+      existingPullRequests,
+    );
+
+    expect(failedUpdates).toBe(0);
+    expect(dependabotCli.update).toHaveBeenCalled();
+    expect(DependabotJobBuilder.updateAllDependenciesJob).toHaveBeenCalled();
+  });
+
+  it('should skip "update all" job if open pull requests limit is reached', async () => {
+    dependabotConfig.updates[0]['open-pull-requests-limit'] = 1;
+    existingPullRequests.push({
+      id: 1,
+      properties: [
+        {
+          name: DependabotOutputProcessor.PR_PROPERTY_NAME_PACKAGE_MANAGER,
+          value: 'npm_and_yarn',
+        },
+        {
+          name: DependabotOutputProcessor.PR_PROPERTY_NAME_DEPENDENCIES,
+          value: JSON.stringify([{ 'dependency-name': 'dependency1' }]),
+        },
+      ],
+    });
+
+    jest.spyOn(tsDependabotOutputProcessor, 'parsePullRequestProperties').mockReturnValue('npm_and_yarn');
+
+    const failedUpdates = await performDependabotUpdatesAsync(
+      taskInputs,
+      dependabotConfig,
+      dependabotConfig.updates,
+      dependabotCli,
+      dependabotCliUpdateOptions,
+      existingPullRequests,
+    );
+
+    expect(failedUpdates).toBe(0);
+    expect(DependabotJobBuilder.updateAllDependenciesJob).not.toHaveBeenCalled();
+  });
+
+  it('should perform "update security-only" job if open pull request limit is zero', async () => {
+    dependabotConfig.updates[0]['open-pull-requests-limit'] = 0;
+    const ghsaClient = new GitHubGraphClient('fake-token');
+    ghsaClient.getSecurityVulnerabilitiesAsync = jest.fn().mockResolvedValue([]);
+
+    const failedUpdates = await performDependabotUpdatesAsync(
+      taskInputs,
+      dependabotConfig,
+      dependabotConfig.updates,
+      dependabotCli,
+      dependabotCliUpdateOptions,
+      existingPullRequests,
+    );
+
+    expect(failedUpdates).toBe(0);
+    expect(DependabotJobBuilder.listAllDependenciesJob).toHaveBeenCalled();
+  });
+
+  it('should perform "update pull request" job successfully if there are existing pull requests', async () => {
+    dependabotConfig.updates[0]['open-pull-requests-limit'] = 1;
+    existingPullRequests.push({
+      id: 1,
+      properties: [
+        {
+          name: DependabotOutputProcessor.PR_PROPERTY_NAME_PACKAGE_MANAGER,
+          value: 'npm_and_yarn',
+        },
+        {
+          name: DependabotOutputProcessor.PR_PROPERTY_NAME_DEPENDENCIES,
+          value: JSON.stringify([{ 'dependency-name': 'dependency1' }]),
+        },
+      ],
+    });
+
+    jest.spyOn(tsDependabotJobBuilder, 'mapPackageEcosystemToPackageManager').mockReturnValue('npm_and_yarn');
+
+    const failedUpdates = await performDependabotUpdatesAsync(
+      taskInputs,
+      dependabotConfig,
+      dependabotConfig.updates,
+      dependabotCli,
+      dependabotCliUpdateOptions,
+      existingPullRequests,
+    );
+
+    expect(failedUpdates).toBe(0);
+    expect(DependabotJobBuilder.updatePullRequestJob).toHaveBeenCalled();
+  });
+});

--- a/extension/tasks/dependabotV2/index.ts
+++ b/extension/tasks/dependabotV2/index.ts
@@ -1,24 +1,24 @@
 import { debug, error, setResult, TaskResult, warning, which } from 'azure-pipelines-task-lib/task';
 import { AzureDevOpsWebApiClient } from './utils/azure-devops/AzureDevOpsWebApiClient';
 import { section, setSecrets } from './utils/azure-devops/formattingCommands';
+import { IPullRequestProperties } from './utils/azure-devops/interfaces/IPullRequest';
 import { DependabotCli } from './utils/dependabot-cli/DependabotCli';
-import { DependabotJobBuilder } from './utils/dependabot-cli/DependabotJobBuilder';
+import { DependabotJobBuilder, mapPackageEcosystemToPackageManager } from './utils/dependabot-cli/DependabotJobBuilder';
 import {
   DependabotOutputProcessor,
   parsePullRequestProperties,
 } from './utils/dependabot-cli/DependabotOutputProcessor';
 import { IDependabotUpdateOperationResult } from './utils/dependabot-cli/interfaces/IDependabotUpdateOperationResult';
-import { IDependabotUpdate } from './utils/dependabot/interfaces/IDependabotConfig';
+import { IDependabotConfig, IDependabotUpdate } from './utils/dependabot/interfaces/IDependabotConfig';
 import parseDependabotConfigFile from './utils/dependabot/parseConfigFile';
-import parseTaskInputConfiguration from './utils/getSharedVariables';
+import parseTaskInputConfiguration, { ISharedVariables } from './utils/getSharedVariables';
 import { GitHubGraphClient } from './utils/github/GitHubGraphClient';
 import { IPackage } from './utils/github/IPackage';
 import { ISecurityVulnerability } from './utils/github/ISecurityVulnerability';
-import { getGhsaPackageEcosystemFromDependabotPackageEcosystem } from './utils/github/PackageEcosystem';
+import { getGhsaPackageEcosystemFromDependabotPackageManager } from './utils/github/PackageEcosystem';
 
 async function run() {
-  let dependabot: DependabotCli = undefined;
-  let failedTasks: number = 0;
+  let dependabotCli: DependabotCli = undefined;
   try {
     // Check if required tools are installed
     debug('Checking for `docker` install...');
@@ -64,12 +64,12 @@ async function run() {
 
     // Initialise the DevOps API clients
     // There are two clients; one for authoring pull requests and one for auto-approving pull requests (if configured)
-    const prAuthorClient = new AzureDevOpsWebApiClient(
+    const devOpsPrAuthorClient = new AzureDevOpsWebApiClient(
       taskInputs.organizationUrl.toString(),
       taskInputs.systemAccessToken,
       taskInputs.debug,
     );
-    const prApproverClient = taskInputs.autoApprove
+    const devOpsPrApproverClient = taskInputs.autoApprove
       ? new AzureDevOpsWebApiClient(
           taskInputs.organizationUrl.toString(),
           taskInputs.autoApproveUserToken || taskInputs.systemAccessToken,
@@ -78,20 +78,20 @@ async function run() {
       : null;
 
     // Fetch the active pull requests created by the author user
-    const existingBranchNames = await prAuthorClient.getBranchNames(taskInputs.project, taskInputs.repository);
-    const existingPullRequests = await prAuthorClient.getActivePullRequestProperties(
+    const existingBranchNames = await devOpsPrAuthorClient.getBranchNames(taskInputs.project, taskInputs.repository);
+    const existingPullRequests = await devOpsPrAuthorClient.getActivePullRequestProperties(
       taskInputs.project,
       taskInputs.repository,
-      await prAuthorClient.getUserId(),
+      await devOpsPrAuthorClient.getUserId(),
     );
 
     // Initialise the Dependabot updater
-    dependabot = new DependabotCli(
+    dependabotCli = new DependabotCli(
       taskInputs.dependabotCliPackage || DependabotCli.CLI_PACKAGE_LATEST,
       new DependabotOutputProcessor(
         taskInputs,
-        prAuthorClient,
-        prApproverClient,
+        devOpsPrAuthorClient,
+        devOpsPrApproverClient,
         existingBranchNames,
         existingPullRequests,
         taskInputs.debug,
@@ -99,7 +99,7 @@ async function run() {
       taskInputs.debug,
     );
 
-    const dependabotUpdaterOptions = {
+    const dependabotCliUpdateOptions = {
       sourceProvider: 'azure',
       azureDevOpsAccessToken: taskInputs.systemAccessToken,
       gitHubAccessToken: taskInputs.githubAccessToken,
@@ -112,144 +112,172 @@ async function run() {
     };
 
     // If update identifiers are specified, select them; otherwise handle all
-    let updates: IDependabotUpdate[] = [];
+    let dependabotUpdatesToPerform: IDependabotUpdate[] = [];
     const targetIds = taskInputs.targetUpdateIds;
     if (targetIds && targetIds.length > 0) {
       for (const id of targetIds) {
-        updates.push(dependabotConfig.updates[id]);
+        dependabotUpdatesToPerform.push(dependabotConfig.updates[id]);
       }
     } else {
-      updates = dependabotConfig.updates;
+      dependabotUpdatesToPerform = dependabotConfig.updates;
     }
 
-    // Loop through the [targeted] update blocks in dependabot.yaml and perform updates
-    for (const update of updates) {
-      const updateId = updates.indexOf(update).toString();
-      const packageEcosystem = update['package-ecosystem'];
-
-      // Parse the Dependabot metadata for the existing pull requests that are related to this update
-      // Dependabot will use this to determine if we need to create new pull requests or update/close existing ones
-      const existingPullRequestsForPackageEcosystem = parsePullRequestProperties(
-        existingPullRequests,
-        packageEcosystem,
-      );
-      const existingPullRequestDependenciesForPackageEcosystem = Object.entries(
-        existingPullRequestsForPackageEcosystem,
-      ).map(([id, deps]) => deps);
-
-      // If this is a security-only update (i.e. 'open-pull-requests-limit: 0'), then we first need to discover the dependencies
-      // that need updating and check each one for vulnerabilities. This is because Dependabot requires the list of vulnerable dependencies
-      // to be supplied in the job definition of security-only update job, it will not automatically discover them like a versioned update does.
-      // https://docs.github.com/en/code-security/dependabot/dependabot-security-updates/configuring-dependabot-security-updates#overriding-the-default-behavior-with-a-configuration-file
-      let securityVulnerabilities: ISecurityVulnerability[] = [];
-      let dependencyNamesToUpdate: string[] = [];
-      const securityUpdatesOnly = update['open-pull-requests-limit'] === 0;
-      if (securityUpdatesOnly) {
-        // Run an update job to discover all dependencies
-        const discoveredDependencyListOutputs = await dependabot.update(
-          DependabotJobBuilder.listAllDependenciesJob(taskInputs, updateId, update, dependabotConfig.registries),
-          dependabotUpdaterOptions,
-        );
-
-        // Get the list of vulnerabilities that apply to the discovered dependencies
-        section(`GHSA dependency vulnerability check`);
-        const ghsaClient = new GitHubGraphClient(taskInputs.githubAccessToken);
-        const packagesToCheckForVulnerabilities: IPackage[] = discoveredDependencyListOutputs
-          ?.find((x) => x.output.type == 'update_dependency_list')
-          ?.output?.data?.dependencies?.map((d) => ({ name: d.name, version: d.version }));
-        if (packagesToCheckForVulnerabilities?.length) {
-          console.info(
-            `Detected ${packagesToCheckForVulnerabilities.length} dependencies; Checking for vulnerabilities...`,
-          );
-          securityVulnerabilities = await ghsaClient.getSecurityVulnerabilitiesAsync(
-            getGhsaPackageEcosystemFromDependabotPackageEcosystem(packageEcosystem),
-            packagesToCheckForVulnerabilities || [],
-          );
-
-          // Only update dependencies that have vulnerabilities
-          dependencyNamesToUpdate = Array.from(new Set(securityVulnerabilities.map((v) => v.package.name)));
-          console.info(
-            `Detected ${securityVulnerabilities.length} vulnerabilities affecting ${dependencyNamesToUpdate.length} dependencies`,
-          );
-          if (dependencyNamesToUpdate.length) {
-            console.log(dependencyNamesToUpdate);
-          }
-        } else {
-          console.info('No vulnerabilities detected in any dependencies');
-        }
-      }
-
-      // Run an update job for "all dependencies"; this will create new pull requests for dependencies that need updating
-      const openPullRequestsLimit = update['open-pull-requests-limit'];
-      const openPullRequestsCount = Object.entries(existingPullRequestsForPackageEcosystem).length;
-      const hasReachedOpenPullRequestLimit =
-        openPullRequestsLimit > 0 && openPullRequestsCount >= openPullRequestsLimit;
-      if (!hasReachedOpenPullRequestLimit) {
-        const dependenciesHaveVulnerabilities = dependencyNamesToUpdate.length && securityVulnerabilities.length;
-        if (!securityUpdatesOnly || dependenciesHaveVulnerabilities) {
-          failedTasks += handleUpdateOperationResults(
-            await dependabot.update(
-              DependabotJobBuilder.updateAllDependenciesJob(
-                taskInputs,
-                updateId,
-                update,
-                dependabotConfig.registries,
-                dependencyNamesToUpdate,
-                existingPullRequestDependenciesForPackageEcosystem,
-                securityVulnerabilities,
-              ),
-              dependabotUpdaterOptions,
-            ),
-          );
-        } else {
-          console.info('Nothing to update; dependencies are not affected by any known vulnerability');
-        }
-      } else {
-        warning(
-          `Skipping update for ${packageEcosystem} packages as the open pull requests limit (${openPullRequestsLimit}) has already been reached`,
-        );
-      }
-
-      // If there are existing pull requests, run an update job for each one; this will resolve merge conflicts and close pull requests that are no longer needed
-      const numberOfPullRequestsToUpdate = Object.keys(existingPullRequestsForPackageEcosystem).length;
-      if (numberOfPullRequestsToUpdate > 0) {
-        if (!taskInputs.skipPullRequests) {
-          for (const pullRequestId in existingPullRequestsForPackageEcosystem) {
-            failedTasks += handleUpdateOperationResults(
-              await dependabot.update(
-                DependabotJobBuilder.updatePullRequestJob(
-                  taskInputs,
-                  pullRequestId,
-                  update,
-                  dependabotConfig.registries,
-                  existingPullRequestDependenciesForPackageEcosystem,
-                  existingPullRequestsForPackageEcosystem[pullRequestId],
-                  securityVulnerabilities,
-                ),
-                dependabotUpdaterOptions,
-              ),
-            );
-          }
-        } else {
-          warning(
-            `Skipping update of ${numberOfPullRequestsToUpdate} existing ${packageEcosystem} package pull request(s) as 'skipPullRequests' is set to 'true'`,
-          );
-        }
-      }
-    }
+    // Perform updates for each of the [targeted] update blocks in dependabot.yaml
+    const failedUpdateOperations = await performDependabotUpdatesAsync(
+      taskInputs,
+      dependabotConfig,
+      dependabotUpdatesToPerform,
+      dependabotCli,
+      dependabotCliUpdateOptions,
+      existingPullRequests,
+    );
 
     setResult(
-      failedTasks ? TaskResult.Failed : TaskResult.Succeeded,
-      failedTasks
-        ? `${failedTasks} update tasks(s) failed, check logs for more information`
+      failedUpdateOperations == 0 ? TaskResult.Succeeded : TaskResult.Failed,
+      failedUpdateOperations > 0
+        ? `${failedUpdateOperations} update tasks(s) failed, check logs for more information`
         : `All update tasks completed successfully`,
     );
   } catch (e) {
     setResult(TaskResult.Failed, e?.message);
     exception(e);
   } finally {
-    dependabot?.cleanup();
+    dependabotCli?.cleanup();
   }
+}
+
+/**
+ * Performs the Dependabot updates.
+ * @param taskInputs The shared task inputs.
+ * @param dependabotConfig The parsed Dependabot configuration.
+ * @param dependabotUpdates The updates to perform.
+ * @param dependabotCli The Dependabot CLI instance.
+ * @param dependabotCliUpdateOptions The Dependabot updater options.
+ * @param existingPullRequests The existing pull requests.
+ * @returns The number of successful and failed update operations.
+ */
+export async function performDependabotUpdatesAsync(
+  taskInputs: ISharedVariables,
+  dependabotConfig: IDependabotConfig,
+  dependabotUpdates: IDependabotUpdate[],
+  dependabotCli: DependabotCli,
+  dependabotCliUpdateOptions: any,
+  existingPullRequests: IPullRequestProperties[],
+): Promise<number> {
+  let failedOperations = 0;
+  for (const update of dependabotUpdates) {
+    const updateId = dependabotUpdates.indexOf(update).toString();
+    const packageEcosystem = update['package-ecosystem'];
+    const packageManager = mapPackageEcosystemToPackageManager(packageEcosystem);
+
+    // Parse the Dependabot metadata for the existing pull requests that are related to this update
+    // Dependabot will use this to determine if we need to create new pull requests or update/close existing ones
+    const existingPullRequestsForPackageManager = parsePullRequestProperties(existingPullRequests, packageManager);
+    const existingPullRequestDependenciesForPackageManager = Object.entries(existingPullRequestsForPackageManager).map(
+      ([id, deps]) => deps,
+    );
+
+    // If this is a security-only update (i.e. 'open-pull-requests-limit: 0'), then we first need to discover the dependencies
+    // that need updating and check each one for vulnerabilities. This is because Dependabot requires the list of vulnerable dependencies
+    // to be supplied in the job definition of security-only update job, it will not automatically discover them like a versioned update does.
+    // https://docs.github.com/en/code-security/dependabot/dependabot-security-updates/configuring-dependabot-security-updates#overriding-the-default-behavior-with-a-configuration-file
+    let securityVulnerabilities: ISecurityVulnerability[] = [];
+    let dependencyNamesToUpdate: string[] = [];
+    const securityUpdatesOnly = update['open-pull-requests-limit'] === 0;
+    if (securityUpdatesOnly) {
+      // Run an update job to discover all dependencies
+      const discoveredDependencyListOutputs = await dependabotCli.update(
+        DependabotJobBuilder.listAllDependenciesJob(taskInputs, updateId, update, dependabotConfig.registries),
+        dependabotCliUpdateOptions,
+      );
+
+      // Get the list of vulnerabilities that apply to the discovered dependencies
+      section(`GHSA dependency vulnerability check`);
+      const ghsaClient = new GitHubGraphClient(taskInputs.githubAccessToken);
+      const packagesToCheckForVulnerabilities: IPackage[] = discoveredDependencyListOutputs
+        ?.find((x) => x.output.type == 'update_dependency_list')
+        ?.output?.data?.dependencies?.map((d) => ({ name: d.name, version: d.version }));
+      if (packagesToCheckForVulnerabilities?.length) {
+        console.info(
+          `Detected ${packagesToCheckForVulnerabilities.length} dependencies; Checking for vulnerabilities...`,
+        );
+        securityVulnerabilities = await ghsaClient.getSecurityVulnerabilitiesAsync(
+          getGhsaPackageEcosystemFromDependabotPackageManager(packageManager),
+          packagesToCheckForVulnerabilities || [],
+        );
+
+        // Only update dependencies that have vulnerabilities
+        dependencyNamesToUpdate = Array.from(new Set(securityVulnerabilities.map((v) => v.package.name)));
+        console.info(
+          `Detected ${securityVulnerabilities.length} vulnerabilities affecting ${dependencyNamesToUpdate.length} dependencies`,
+        );
+        if (dependencyNamesToUpdate.length) {
+          console.log(dependencyNamesToUpdate);
+        }
+      } else {
+        console.info('No vulnerabilities detected in any dependencies');
+      }
+    }
+
+    // Run an update job for "all dependencies"; this will create new pull requests for dependencies that need updating
+    const openPullRequestsLimit = update['open-pull-requests-limit'];
+    const openPullRequestsCount = Object.entries(existingPullRequestsForPackageManager).length;
+    const hasReachedOpenPullRequestLimit = openPullRequestsLimit > 0 && openPullRequestsCount >= openPullRequestsLimit;
+    if (!hasReachedOpenPullRequestLimit) {
+      const dependenciesHaveVulnerabilities = dependencyNamesToUpdate.length && securityVulnerabilities.length;
+      if (!securityUpdatesOnly || dependenciesHaveVulnerabilities) {
+        failedOperations += handleUpdateOperationResults(
+          await dependabotCli.update(
+            DependabotJobBuilder.updateAllDependenciesJob(
+              taskInputs,
+              updateId,
+              update,
+              dependabotConfig.registries,
+              dependencyNamesToUpdate,
+              existingPullRequestDependenciesForPackageManager,
+              securityVulnerabilities,
+            ),
+            dependabotCliUpdateOptions,
+          ),
+        );
+      } else {
+        console.info('Nothing to update; dependencies are not affected by any known vulnerability');
+      }
+    } else {
+      warning(
+        `Skipping update for ${packageEcosystem} packages as the open pull requests limit (${openPullRequestsLimit}) has already been reached`,
+      );
+    }
+
+    // If there are existing pull requests, run an update job for each one; this will resolve merge conflicts and close pull requests that are no longer needed
+    const numberOfPullRequestsToUpdate = Object.keys(existingPullRequestsForPackageManager).length;
+    if (numberOfPullRequestsToUpdate > 0) {
+      if (!taskInputs.skipPullRequests) {
+        for (const pullRequestId in existingPullRequestsForPackageManager) {
+          failedOperations += handleUpdateOperationResults(
+            await dependabotCli.update(
+              DependabotJobBuilder.updatePullRequestJob(
+                taskInputs,
+                pullRequestId,
+                update,
+                dependabotConfig.registries,
+                existingPullRequestDependenciesForPackageManager,
+                existingPullRequestsForPackageManager[pullRequestId],
+                securityVulnerabilities,
+              ),
+              dependabotCliUpdateOptions,
+            ),
+          );
+        }
+      } else {
+        warning(
+          `Skipping update of ${numberOfPullRequestsToUpdate} existing ${packageEcosystem} package pull request(s) as 'skipPullRequests' is set to 'true'`,
+        );
+      }
+    }
+  }
+
+  return failedOperations;
 }
 
 /**
@@ -258,25 +286,24 @@ async function run() {
  * @returns The number of failed tasks (i.e. outputs that could not be processed).
  * @remarks
  * If the update operation completed with all outputs processed successfully, it will return 0.
- * If the update operation completed with no outputs, it will return 1.
  * If the update operation completed with some outputs processed unsuccessfully, it will return the number of failed outputs.
  */
-function handleUpdateOperationResults(outputs: IDependabotUpdateOperationResult[] | undefined) {
-  let failedTasks = 0; // assume success, initially
+function handleUpdateOperationResults(outputs: IDependabotUpdateOperationResult[] | undefined): number {
+  let failedOperations = 0; // assume success, initially
   if (outputs) {
     // The update operation completed, but some output tasks may have failed
-    const failedUpdateTasks = outputs.filter((u) => !u.success);
-    if (failedUpdateTasks.length > 0) {
+    const failedUpdateOutputs = outputs.filter((u) => !u.success);
+    if (failedUpdateOutputs.length > 0) {
       // At least one output task failed to process
-      failedUpdateTasks.forEach((u) => exception(u.error));
-      failedTasks += failedUpdateTasks.length;
+      failedUpdateOutputs.forEach((u) => exception(u.error));
+      failedOperations += failedUpdateOutputs.length;
     }
   } else {
     // The update operation critically failed, it produced no output
-    failedTasks++;
+    failedOperations++;
   }
 
-  return failedTasks;
+  return failedOperations;
 }
 
 function exception(e: Error) {

--- a/extension/tasks/dependabotV2/utils/azure-devops/AzureDevOpsWebApiClient.test.ts
+++ b/extension/tasks/dependabotV2/utils/azure-devops/AzureDevOpsWebApiClient.test.ts
@@ -16,9 +16,6 @@ describe('AzureDevOpsWebApiClient', () => {
 
   beforeEach(() => {
     client = new AzureDevOpsWebApiClient(organisationApiUrl, accessToken);
-  });
-
-  afterEach(() => {
     jest.clearAllMocks();
   });
 

--- a/extension/tasks/dependabotV2/utils/dependabot-cli/DependabotJobBuilder.test.ts
+++ b/extension/tasks/dependabotV2/utils/dependabot-cli/DependabotJobBuilder.test.ts
@@ -14,40 +14,21 @@ import {
 
 describe('mapSourceFromDependabotConfigToJobConfig', () => {
   it('should map source correctly for Azure DevOps Services', () => {
-    const taskInputs: ISharedVariables = {
+    const taskInputs = {
       apiEndpointUrl: 'https://dev.azure.com',
       hostname: 'dev.azure.com',
       organization: 'my-org',
       project: 'my-project',
       repository: 'my-repo',
       virtualDirectory: '',
-      port: '',
-      organizationUrl: undefined,
-      protocol: '',
-      projectId: '',
-      repositoryOverridden: false,
-      githubAccessToken: '',
-      systemAccessUser: '',
-      systemAccessToken: '',
-      storeDependencyList: false,
-      setAutoComplete: false,
-      mergeStrategy: '',
-      autoCompleteIgnoreConfigIds: [],
-      autoApprove: false,
-      autoApproveUserToken: '',
-      experiments: undefined,
-      debug: false,
-      targetUpdateIds: [],
-      securityAdvisoriesFile: '',
-      skipPullRequests: false,
-      commentPullRequests: false,
-      abandonUnwantedPullRequests: false,
-    };
-    const update: IDependabotUpdate = {
+      port: '443',
+      protocol: 'https',
+    } as ISharedVariables;
+    const update = {
       'package-ecosystem': 'nuget',
       'directory': '/',
       'directories': [],
-    };
+    } as IDependabotUpdate;
 
     const result = mapSourceFromDependabotConfigToJobConfig(taskInputs, update);
     expect(result).toMatchObject({
@@ -59,7 +40,7 @@ describe('mapSourceFromDependabotConfigToJobConfig', () => {
   });
 
   it('should map source correctly for Azure DevOps Server', () => {
-    const taskInputs: ISharedVariables = {
+    const taskInputs = {
       apiEndpointUrl: 'https://my-org.com:8443/tfs',
       hostname: 'my-org.com',
       organization: 'my-collection',
@@ -67,32 +48,13 @@ describe('mapSourceFromDependabotConfigToJobConfig', () => {
       repository: 'my-repo',
       virtualDirectory: 'tfs',
       port: '8443',
-      organizationUrl: undefined,
-      protocol: '',
-      projectId: '',
-      repositoryOverridden: false,
-      githubAccessToken: '',
-      systemAccessUser: '',
-      systemAccessToken: '',
-      storeDependencyList: false,
-      setAutoComplete: false,
-      mergeStrategy: '',
-      autoCompleteIgnoreConfigIds: [],
-      autoApprove: false,
-      autoApproveUserToken: '',
-      experiments: undefined,
-      debug: false,
-      targetUpdateIds: [],
-      securityAdvisoriesFile: '',
-      skipPullRequests: false,
-      commentPullRequests: false,
-      abandonUnwantedPullRequests: false,
-    };
-    const update: IDependabotUpdate = {
+      protocol: 'https',
+    } as ISharedVariables;
+    const update = {
       'package-ecosystem': 'nuget',
       'directory': '/',
       'directories': [],
-    };
+    } as IDependabotUpdate;
 
     const result = mapSourceFromDependabotConfigToJobConfig(taskInputs, update);
     expect(result).toMatchObject({

--- a/extension/tasks/dependabotV2/utils/dependabot-cli/DependabotOutputProcessor.test.ts
+++ b/extension/tasks/dependabotV2/utils/dependabot-cli/DependabotOutputProcessor.test.ts
@@ -39,16 +39,13 @@ describe('DependabotOutputProcessor', () => {
     let data: any;
 
     beforeEach(() => {
+      jest.clearAllMocks();
       update = {
         job: {} as any,
         credentials: {} as any,
         config: {} as IDependabotUpdate,
       } as IDependabotUpdateOperation;
       data = {};
-    });
-
-    afterEach(() => {
-      jest.clearAllMocks();
     });
 
     it('should skip processing "update_dependency_list" if "storeDependencyList" is false', async () => {
@@ -58,7 +55,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'update_dependency_list', data);
 
       expect(result).toBe(true);
-      expect(prApproverClient.updateProjectProperty).toHaveBeenCalledTimes(0);
+      expect(prApproverClient.updateProjectProperty).not.toHaveBeenCalled();
     });
 
     it('should process "update_dependency_list"', async () => {
@@ -68,7 +65,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'update_dependency_list', data);
 
       expect(result).toBe(true);
-      expect(prAuthorClient.updateProjectProperty).toHaveBeenCalledTimes(1);
+      expect(prAuthorClient.updateProjectProperty).toHaveBeenCalled();
     });
 
     it('should skip processing "create_pull_request" if "skipPullRequests" is true', async () => {
@@ -77,7 +74,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'create_pull_request', data);
 
       expect(result).toBe(true);
-      expect(prAuthorClient.createPullRequest).toHaveBeenCalledTimes(0);
+      expect(prAuthorClient.createPullRequest).not.toHaveBeenCalled();
     });
 
     it('should skip processing "create_pull_request" if open pull request limit is reached', async () => {
@@ -86,7 +83,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'create_pull_request', data);
 
       expect(result).toBe(true);
-      expect(prAuthorClient.createPullRequest).toHaveBeenCalledTimes(0);
+      expect(prAuthorClient.createPullRequest).not.toHaveBeenCalled();
     });
 
     it('should process "create_pull_request"', async () => {
@@ -107,8 +104,8 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'create_pull_request', data);
 
       expect(result).toBe(true);
-      expect(prAuthorClient.createPullRequest).toHaveBeenCalledTimes(1);
-      expect(prApproverClient.approvePullRequest).toHaveBeenCalledTimes(1);
+      expect(prAuthorClient.createPullRequest).toHaveBeenCalled();
+      expect(prApproverClient.approvePullRequest).toHaveBeenCalled();
     });
 
     it('should skip processing "update_pull_request" if "skipPullRequests" is false', async () => {
@@ -117,7 +114,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'update_pull_request', data);
 
       expect(result).toBe(true);
-      expect(prAuthorClient.updatePullRequest).toHaveBeenCalledTimes(0);
+      expect(prAuthorClient.updatePullRequest).not.toHaveBeenCalled();
     });
 
     it('should fail processing "update_pull_request" if pull request does not exist', async () => {
@@ -128,7 +125,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'update_pull_request', data);
 
       expect(result).toBe(false);
-      expect(prAuthorClient.updatePullRequest).toHaveBeenCalledTimes(0);
+      expect(prAuthorClient.updatePullRequest).not.toHaveBeenCalled();
     });
 
     it('should process "update_pull_request"', async () => {
@@ -158,8 +155,8 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'update_pull_request', data);
 
       expect(result).toBe(true);
-      expect(prAuthorClient.updatePullRequest).toHaveBeenCalledTimes(1);
-      expect(prApproverClient.approvePullRequest).toHaveBeenCalledTimes(1);
+      expect(prAuthorClient.updatePullRequest).toHaveBeenCalled();
+      expect(prApproverClient.approvePullRequest).toHaveBeenCalled();
     });
 
     it('should skip processing "close_pull_request" if "abandonUnwantedPullRequests" is false', async () => {
@@ -168,7 +165,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'close_pull_request', data);
 
       expect(result).toBe(true);
-      expect(prAuthorClient.abandonPullRequest).toHaveBeenCalledTimes(0);
+      expect(prAuthorClient.abandonPullRequest).not.toHaveBeenCalled();
     });
 
     it('should fail processing "close_pull_request" if pull request does not exist', async () => {
@@ -180,7 +177,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'close_pull_request', data);
 
       expect(result).toBe(false);
-      expect(prAuthorClient.abandonPullRequest).toHaveBeenCalledTimes(0);
+      expect(prAuthorClient.abandonPullRequest).not.toHaveBeenCalled();
     });
 
     it('should process "close_pull_request"', async () => {
@@ -205,7 +202,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'close_pull_request', data);
 
       expect(result).toBe(true);
-      expect(prAuthorClient.abandonPullRequest).toHaveBeenCalledTimes(1);
+      expect(prAuthorClient.abandonPullRequest).toHaveBeenCalled();
     });
 
     it('should process "mark_as_processed"', async () => {
@@ -224,14 +221,14 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'record_update_job_error', data);
 
       expect(result).toBe(false);
-      expect(error).toHaveBeenCalledTimes(1);
+      expect(error).toHaveBeenCalled();
     });
 
     it('should process "record_update_job_unknown_error"', async () => {
       const result = await processor.process(update, 'record_update_job_unknown_error', data);
 
       expect(result).toBe(false);
-      expect(error).toHaveBeenCalledTimes(1);
+      expect(error).toHaveBeenCalled();
     });
 
     it('should process "increment_metric"', async () => {
@@ -244,7 +241,7 @@ describe('DependabotOutputProcessor', () => {
       const result = await processor.process(update, 'non_existant_output_type', data);
 
       expect(result).toBe(true);
-      expect(warning).toHaveBeenCalledTimes(1);
+      expect(warning).toHaveBeenCalled();
     });
   });
 });

--- a/extension/tasks/dependabotV2/utils/github/PackageEcosystem.ts
+++ b/extension/tasks/dependabotV2/utils/github/PackageEcosystem.ts
@@ -13,10 +13,10 @@ export enum PackageEcosystem {
   Swift = 'SWIFT',
 }
 
-export function getGhsaPackageEcosystemFromDependabotPackageEcosystem(
-  dependabotPackageEcosystem: string,
+export function getGhsaPackageEcosystemFromDependabotPackageManager(
+  dependabotPackageManager: string,
 ): PackageEcosystem {
-  switch (dependabotPackageEcosystem) {
+  switch (dependabotPackageManager) {
     case 'composer':
       return PackageEcosystem.Composer;
     case 'elm':
@@ -42,6 +42,6 @@ export function getGhsaPackageEcosystemFromDependabotPackageEcosystem(
     case 'swift':
       return PackageEcosystem.Swift;
     default:
-      throw new Error(`Unknown dependabot package ecosystem: ${dependabotPackageEcosystem}`);
+      throw new Error(`Unknown dependabot package manager: ${dependabotPackageManager}`);
   }
 }

